app/controllers/components_controller.rb

```diff
@@ -2,15 +2,19 @@
 
 class ComponentsController < ApplicationController
   def show
-    @component = Component.get(params[:id])
+    @component = Component.get(params[:id]) || not_found
   end
 
   def index
     @components = Component.all
   end
 
   def preview
-    @component = Component.get(params[:component_id])
+    @component = Component.get(params[:component_id]) || not_found
     render layout: 'preview'
   end
+
+  def not_found
+    render file: "#{Rails.root}/public/404", layout: false, status: :not_found
```

Comment:
Rails has a `config.exceptions_app` configuration that defaults to [ActionDispatch::PublicExceptions](http://api.rubyonrails.org/classes/ActionDispatch/PublicExceptions.html) and will take care of the file rendering.

All you need to do is `render :not_found` if the component is nil.